arwen/host/execution_async_context.go

```diff
@@ -0,0 +1,435 @@
+	output := host.Output()
+
+	err := output.Transfer(
+		asyncCall.GetDestination(),
+		runtime.GetSCAddress(),
+		asyncCall.GetGasLimit(),
+		asyncCall.GetGasLocked(),
+		big.NewInt(0).SetBytes(asyncCall.GetValueBytes()),
+		asyncCall.GetData(),
+		vmcommon.AsynchronousCall,
+	)
+	if err != nil {
+		metering := host.Metering()
+		metering.UseGas(metering.GasLeft())
+		runtime.FailExecution(err)
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * setupAsyncCallsGas sets the gasLimit for each async call with the amount of gas provided by the
+ *  SC developer. The remaining gas is split between the async calls where the developer
+ *  did not specify any gas amount
+ */
+func (host *vmHost) setupAsyncCallsGas(asyncContext *arwen.AsyncContext) error {
+	gasLeft := host.Metering().GasLeft()
+	gasNeeded := uint64(0)
+	callsWithZeroGas := uint64(0)
+
+	for _, asyncCallGroup := range asyncContext.AsyncCallGroups {
+		for _, asyncCall := range asyncCallGroup.AsyncCalls {
+			var err error
+			gasNeeded, err = math.AddUint64(gasNeeded, asyncCall.ProvidedGas)
+			if err != nil {
+				return err
+			}
+
+			if gasNeeded > gasLeft {
+				return arwen.ErrNotEnoughGas
+			}
+
+			if asyncCall.ProvidedGas == 0 {
+				callsWithZeroGas++
+				continue
+			}
+
+			asyncCall.GasLimit = asyncCall.ProvidedGas
+		}
+	}
+
+	if callsWithZeroGas == 0 {
+		return nil
+	}
+
+	if gasLeft <= gasNeeded {
+		return arwen.ErrNotEnoughGas
+	}
+
+	gasShare := (gasLeft - gasNeeded) / callsWithZeroGas
+	for _, asyncCallGroup := range asyncContext.AsyncCallGroups {
+		for _, asyncCall := range asyncCallGroup.AsyncCalls {
+			if asyncCall.ProvidedGas == 0 {
+				asyncCall.GasLimit = gasShare
+			}
+		}
+	}
+
+	return nil
+}
+
+func (host *vmHost) finishSyncExecution(vmOutput *vmcommon.VMOutput, err error) {
+	if err == nil {
+		return
+	}
+
+	runtime := host.Runtime()
+	output := host.Output()
+
+	runtime.GetVMInput().GasProvided = 0
+
+	if vmOutput == nil {
+		vmOutput = output.CreateVMOutputInCaseOfError(err)
+	}
+
+	output.SetReturnMessage(vmOutput.ReturnMessage)
+
+	output.Finish([]byte(vmOutput.ReturnCode.String()))
+	output.Finish(runtime.GetCurrentTxHash())
+}
+
+/**
+ * saveAsyncContext takes a list of pending async calls and save them to storage so the info will be available on callback
+ */
+func (host *vmHost) saveAsyncContext(asyncContext *arwen.AsyncContext) error {
+	if len(asyncContext.AsyncCallGroups) == 0 {
+		return nil
+	}
+
+	storage := host.Storage()
+	runtime := host.Runtime()
+
+	asyncCallStorageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	data, err := json.Marshal(asyncContext)
+	if err != nil {
+		return err
+	}
+
+	_, err = storage.SetStorage(asyncCallStorageKey, data)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (host *vmHost) computeDataLengthFromArguments(function string, arguments [][]byte) int {
+	// Calculate what length would the Data field have, were it of the
+	// form "callback@arg1@arg4...
```

Comment:
no need to compute like that. those costs are amortized in other ways.

Reply:
This can be simplified, but it's already done - would leave it like this for now, unless there's something wrong with it.